src/Uno.UI/Controls/NativeFramePresenter.iOS.cs

```diff
@@ -139,40 +381,290 @@ private bool GetIsAnimated(PageStackEntry entry)
+				// It's possibly a modal ViewController.
+				// We just ignore it.
+			}
+		}
+
+		private static bool TryGetFirst(LinkedList<NavigationRequest> navigationRequests, out NavigationRequest firstValue)
+		{
+			firstValue = navigationRequests.FirstOrDefault();
+			return firstValue != null;
+		}
 
-			// Here, we detect whether the newly pushed viewController was triggered by:
-			// - Frame events (the viewController will be part of the _requestedViewControllers queue)
-			// - UINavigationController user interactions (e.g., back button, back swipe)
+		private static bool TryGetLast(LinkedList<NavigationRequest> navigationRequests, out NavigationRequest lastValue)
+		{
+			lastValue = navigationRequests.LastOrDefault();
+			return lastValue != null;
+		}
 
-			// Iterate through the _requestedViewControllers queue to find the newly pushed viewController
-			while (_requestedViewControllers.Any())
+		/// <summary>
+		/// This represents a navigation request.
+		/// We can create this object from <see cref="Frame"/> events (<see cref="Frame.Navigating"/> & <see cref="Frame.Navigated"/>)
+		/// or from <see cref="UINavigationController"/> events (<see cref="ControllerDelegate.WillShowViewController(UINavigationController, UIViewController, bool)"/> & <see cref="ControllerDelegate.WillShowViewController(UINavigationController, UIViewController, bool)"/>).
+		/// We use this class to correlate requests from the frame with requests from the navigation controller.
+		/// </summary>
+		private class NavigationRequest
+		{
+			/// <summary>
+			/// Constructor for the <see cref="Frame.Navigating"/> event.
+			/// </summary>
+			public NavigationRequest(Frame frame, NavigatingCancelEventArgs e)
 			{
-				if (_requestedViewControllers.Peek() == viewController)
+				NavigationMode = e.NavigationMode;
+				PageType = e.SourcePageType;
+				TransitionInfo = e.NavigationTransitionInfo;
+
+				// Here we build the BackStack that we would have after the Navigated event.
+				switch (e.NavigationMode)
 				{
-					// The viewController is next in queue, everything is in order.
-					// We keep the viewController in the queue (Peek) in case DidShowViewController is called twice 
-					// with the same viewController (this can happen when dismissing a modal view controller).
-					return;
+					case NavigationMode.New:
+					case NavigationMode.Forward:
+						var backStackPageType = frame.BackStack
+							.Select(p => p.SourcePageType)
+							.ToList();
+						if (frame.Content is Page page)
+						{
+							backStackPageType.Add(page.GetType());
+						}
+						BackStackPageTypes = backStackPageType;
+						break;
+					case NavigationMode.Back:
+						BackStackPageTypes = frame.BackStack
+							.Take(Math.Max(0, frame.BackStack.Count - 1))
+							.Select(p => p.SourcePageType)
+							.ToList();
+						break;
+					case NavigationMode.Refresh:
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this empty 'case' clause.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5436574)